File: Day3/find_the_kth_character_in_string_game_i.java
class Solution {
    public char kthCharacter(int k) {
        StringBuilder prevString = new StringBuilder("a");
        while (prevString.length() < k) {
            StringBuilder currString = new StringBuilder();
            for (int i = 0; i < prevString.length(); i++) {
                char ch = prevString.charAt(i);
                char nextChar = (char) ('a' + ((ch - 'a' + 1) % 26));
                currString.append(nextChar);
            }
            prevString.append(currString);
        }
        return prevString.charAt(k - 1);
    }
}



// Logarithmic approach

class Solution {
    public char kthCharacter(int k) {
        int ans = 0;
        while (k != 1) {
            int power = 31 - Integer.numberOfLeadingZeros(k); // bit_length - 1
            if ((1 << power) == k) power--;
            k -= 1 << power;
            ans++;
        }
        return (char) ('a' + ans);
    }
}



// Bit Count (O(1))

class Solution {
    public char kthCharacter(int k) {
        int bitCount = Integer.bitCount(k - 1);
        return (char) ('a' + bitCount);
    }
}

File: Day3/find_the_kth_character_in_string_game_i.js
class Solution {
    kthCharacter(k) {
        let prevString = 'a';
        while (prevString.length < k) {
            let currString = '';
            for (let ch of prevString) {
                let nextChar = String.fromCharCode(((ch.charCodeAt(0) - 97 + 1) % 26) + 97);
                currString += nextChar;
            }
            prevString += currString;
        }
        return prevString[k - 1];
    }
}


// Logarithmic approach

class Solution {
    kthCharacter(k) {
        let ans = 0;
        while (k !== 1) {
            let power = Math.floor(Math.log2(k));
            if ((1 << power) === k) power -= 1;
            k -= 1 << power;
            ans++;
        }
        return String.fromCharCode('a'.charCodeAt(0) + ans);
    }
}


// Bit Count (O(1))
class Solution {
    kthCharacter(k) {
        const bitCount = (k - 1).toString(2).split('1').length - 1;
        return String.fromCharCode('a'.charCodeAt(0) + bitCount);
    }
}
